lib/src/rules/unused_top_members_in_executable_libraries.dart

```diff
@@ -0,0 +1,166 @@
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  _Visitor(this.rule);
+
+  final LintRule rule;
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    // TODO(a14n): add support of libs with parts
+    if (node.directives.whereType<PartOfDirective>().isNotEmpty) return;
+    if (node.directives.whereType<PartDirective>().isNotEmpty) return;
+
+    var topDeclarations = node.declarations
+        .expand((e) => [
+              if (e is TopLevelVariableDeclaration)
+                ...e.variables.variables
+              else
+                e,
+            ])
+        .toSet();
+
+    if (!topDeclarations.any(_isEntryPoint)) return;
+
+    var declarationByElement = <Element, Declaration>{};
+    for (var declaration in topDeclarations) {
+      var element = declaration.declaredElement;
+      if (element != null) {
+        if (element is TopLevelVariableElement) {
+          declarationByElement[element] = declaration;
+          var getter = element.getter;
+          if (getter != null) declarationByElement[getter] = declaration;
+          var setter = element.setter;
+          if (setter != null) declarationByElement[setter] = declaration;
+        } else {
+          declarationByElement[element] = declaration;
+        }
+      }
+    }
+
+    var dependencies = Map<Declaration, Set<Declaration>>.fromIterable(
+      topDeclarations,
+      value: (e) => DartTypeUtilities.traverseNodesInDFS(e as Declaration)
+          .expand((e) => [
+                if (e is SimpleIdentifier) e.staticElement,
+                // with `id++` staticElement of `id` is null
+                if (e is CompoundAssignmentExpression) ...[
+                  e.readElement,
+                  e.writeElement,
+                ],
+              ])
+          .whereNotNull()
+          .map((e) => declarationByElement[e])
+          .whereNotNull()
+          .toSet(),
+    );
+
+    var entryPoints = topDeclarations.where(_isEntryPoint);
+    if (entryPoints.isEmpty) return;
+
+    var usedMembers = entryPoints.toSet();
+    var toTraverse = usedMembers.toList();
+    while (toTraverse.isNotEmpty) {
+      var declaration = toTraverse.removeLast();
+      for (var dep in dependencies[declaration]!) {
+        if (usedMembers.add(dep)) {
+          toTraverse.add(dep);
+        }
+      }
+    }
+
+    var unusedMembers = topDeclarations.difference(usedMembers).where((e) {
+      var element = e.declaredElement;
+      return element != null &&
+          element.isPublic &&
+          !element.hasVisibleForTesting;
+    });
+    unusedMembers.forEach(rule.reportLint);
+  }
+
+  bool _isEntryPoint(Declaration e) =>
+      e is FunctionDeclaration &&
+      (e.name.name == 'main' || e.metadata.any(_isPragmaVmEntry));
+
+  bool _isPragmaVmEntry(Annotation annotation) {
+    var elementAnnotation = annotation.elementAnnotation;
+    if (elementAnnotation != null) {
+      var value = elementAnnotation.computeConstantValue();
```

Comment:
> we can perhaps delay this if we know that `elementAnnotation.element` is not the `pragma` class?

`elementAnnotation.element` would point to the ConstructorElement and it will work most of the time. Unfortunatelly I think this wouldn't support code like:

```dart
const entryPoint = pragma('vm:entry-point');

@entryPoint
f(){}
```